x11rb-protocol/src/connection/mod.rs

```diff
@@ -318,6 +349,63 @@ impl Connection {
             .pop_front()
             .map(|(seqno, event)| (event, seqno))
     }
+
+    /// Indicate to the connection that it should search for the given special
+    /// event, with the given extension code and extension ID.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the queue ID is not unique.
+    pub fn create_special_event_queue(&mut self, extcode: u8, eid: u32) {
+        if self
+            .special_event_queues
+            .insert(eid, VecDeque::new())
+            .is_some()
+        {
+            panic!("Special event queue IDs should be unique.")
+        }
+        self.classifier.add_exteid_class(extcode, eid);
+    }
+
+    /// Indicate to the connection that it should classify events that fulfill
+    /// the callback as special events.
+    ///
+    /// This should use a unique EID for identification purposes.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the queue ID is not unique.
+    pub fn create_callback_special_event_queue(
+        &mut self,
+        eid: u32,
+        callback: impl FnMut(&[u8]) -> bool + Send + Sync + 'static,
+    ) {
+        if self
+            .special_event_queues
+            .insert(eid, VecDeque::new())
+            .is_some()
+        {
+            panic!("Special event queue IDs should be unique.")
+        }
+        self.classifier.add_callback_class(eid, Box::new(callback));
+    }
```

Comment:
So, if I understood this well:
* For `create_special_event_queue`, `eid` is both the key in `special_event_queues` and a specific field found in events of the present extension (so the key is related to the condition)
* For `create_callback_special_event_queue`, `eid` is just the key in `special_event_queues` (so the key is unrelated to the condition)

I find it a bit confusing, because for the first function, the hash map key (`eid`) is related to the event filter condition, while for the second function, they are completely independent.

What about just keeping `create_callback_special_event_queue`? In my opinion, the present extension case is too specific.

Reply:
That works by me. I’ll change it to be the callback only.